### PR TITLE
Support web sockets in async-http-client

### PR DESCRIPTION
**Pull request by rossabaker**

### CODE REVIEW

File: async-http-client/src/main/scala/org/http4s/client/asynchttpclient/Http4sWebSocketListener.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    extends DefaultWebSocketListener {
+  private[this] val log = getLogger
+
+  val src = boundedQueue[WebSocketFrame](10)
+  var ahcWs: AhcWebSocket = _
+
+  override def onOpen(ws: AhcWebSocket): Unit = {
+    ahcWs = ws
+    val sink: Sink[Task, WebSocketFrame] = Process.constant {
+      case Text(str, _) =>
```
Comment (rossabaker) [inline async-http-client/src/main/scala/org/http4s/client/asynchttpclient/Http4sWebSocketListener.scala]: Needs to support streaming

File: async-http-client/src/main/scala/org/http4s/client/asynchttpclient/Http4sWebSocketListener.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        Task.delay {
+          ahcWs.sendMessage(str.getBytes)
+        }
+      case Binary(data, _) =>
```
Comment (rossabaker) [inline async-http-client/src/main/scala/org/http4s/client/asynchttpclient/Http4sWebSocketListener.scala]: Needs to support streaming

File: async-http-client/src/main/scala/org/http4s/client/asynchttpclient/Http4sWebSocketListener.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        Task.delay {
+          ahcWs.sendMessage(data)
+        }
+      case Close(_) =>
+        Task.delay {
+          ahcWs.close()
```
Comment (rossabaker) [inline async-http-client/src/main/scala/org/http4s/client/asynchttpclient/Http4sWebSocketListener.scala]: How do we close with a code and reason?  We're throwing things away here.

File: async-http-client/src/main/scala/org/http4s/client/asynchttpclient/Http4sWebSocketListener.scala [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    log.info(s"WebSocket closed: ${ws.getLocalAddress} to ${ws.getRemoteAddress}")
+    src.close.run
+  }
+
+  override def onError(t: Throwable): Unit =
+    t match {
+      case _: Terminated =>
+      case _ =>
+        // TODO How can we return a full Response here?
+        // All we get if the status != 101 is an IllegalStateException
```
Comment (rossabaker) [inline async-http-client/src/main/scala/org/http4s/client/asynchttpclient/Http4sWebSocketListener.scala]: I think we need an upstream fix.  The best we can do is return an empty response with the status code, and lose all the headers and body.

**Comment by rossabaker** [comment]
This isn't totally mature yet, but I think it's far enough along to be worth discussing, especially while we're thinking about how to support this in blaze-client.

**Comment by bryce-anderson** [comment]
Sorry for not looking at this for a while. I think the API is on the right track. One thing I don't like about the Stream focused API is that it makes it a bit more awkward (IMO) to make endpoints where the input and output are dependent on one another and not in a trivial way such as mapping.

This is the PR that initially got me on the 'why not just a `Task[A]` and `A => Task[Unit]` soapbox.

**Pull request closed without merging**